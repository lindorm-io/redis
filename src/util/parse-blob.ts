import { TObject, isObject } from "@lindorm-io/core";
import { isArray, isBoolean, isObject as _isObject, isString } from "lodash";

const parseErrorValue = (input: any): Error => {
  const parsed = _isObject(input) ? input : JSON.parse(input);

  const error: any = new Error(parsed.message);
  error.stack = parsed.stack;

  for (const [key, value] of Object.entries(parsed)) {
    if (key === "message" || key === "stack") continue;
    error[key] = value;
  }

  return error;
};

const parseArrayValue = (input: any, meta: TObject<any>): Array<any> => {
  const parsed = isArray(input) ? input : JSON.parse(input);

  const result: Array<any> = [];

  for (const [index, value] of parsed.entries()) {
    if (isObject(meta[index])) {
      result.push(parseObjectValue(value, meta[index]));
    } else if (isArray(meta[index])) {
      result.push(parseArrayValue(value, meta[index]));
    } else if (meta[index] === "boolean") {
      result.push(isBoolean(value) ? value : JSON.parse(value));
    } else if (meta[index] === "date") {
      result.push(new Date(value));
    } else if (meta[index] === "error") {
      result.push(parseErrorValue(value));
    } else if (meta[index] === "number") {
      result.push(parseInt(value, 10));
    } else if (meta[index] === "string") {
      result.push(value);
    } else if (meta[index] === "null") {
      result.push(null);
    } else if (meta[index] === "undefined") {
      result.push(undefined);
    }
  }

  return result;
};

const parseObjectValue = (json: TObject<any>, meta: TObject<any>): TObject<any> => {
  const result: TObject<any> = {};

  for (const [key, value] of Object.entries(json)) {
    if (isObject(meta[key])) {
      result[key] = parseObjectValue(json[key], meta[key]);
    } else if (isArray(meta[key])) {
      result[key] = parseArrayValue(json[key], meta[key]);
    } else if (meta[key] === "boolean") {
      result[key] = JSON.parse(value);
    } else if (meta[key] === "date") {
      result[key] = new Date(value);
    } else if (meta[key] === "error") {
      result[key] = parseErrorValue(value);
    } else if (meta[key] === "number") {
      result[key] = parseInt(value, 10);
    } else if (meta[key] === "string") {
      result[key] = isString(value) ? value : JSON.parse(value);
    } else if (meta[key] === "null") {
      result[key] = null;
    } else if (meta[key] === "undefined") {
      result[key] = undefined;
    }
  }

  return result;
};

export const parseBlob = (string: string): TObject<any> => {
  const { json, meta } = JSON.parse(string);
  return parseObjectValue(json, meta);
};
