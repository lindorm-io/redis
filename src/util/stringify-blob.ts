import { isObject } from "@lindorm-io/core";
import { isArray, isBoolean, isDate, isError, isNull, isNumber, isString, isUndefined } from "lodash";

export const getType = (value: any): string => {
  if (isArray(value)) return "array";
  if (isBoolean(value)) return "boolean";
  if (isDate(value)) return "date";
  if (isError(value)) return "error";
  if (isNull(value)) return "null";
  if (isNumber(value)) return "number";
  if (isString(value)) return "string";
  if (isUndefined(value)) return "undefined";
};

export const getMetaArray = (input: Array<any>): Array<string> => {
  const result: Array<any> = [];

  for (const value of input) {
    if (isObject(value)) {
      result.push(getMetaObject(value));
    } else if (isArray(value)) {
      result.push(getMetaArray(value));
    } else {
      result.push(getType(value));
    }
  }

  return result;
};

export const getMetaObject = (input: Record<string, any>): Record<string, string> => {
  const result: Record<string, any> = {};

  for (const [key, value] of Object.entries(input)) {
    if (isObject(value)) {
      result[key] = getMetaObject(value);
    } else if (isArray(value)) {
      result[key] = getMetaArray(value);
    } else {
      result[key] = getType(value);
    }
  }

  return result;
};

const stringifyArrayValues = (input: Array<any>): Array<any> => {
  const result: Array<any> = [];

  for (const value of input) {
    if (isObject(value)) {
      result.push(stringifyObjectValues(value));
    } else if (isArray(value)) {
      result.push(stringifyArrayValues(value));
    } else if (isDate(value)) {
      result.push(value.toJSON());
    } else if (isError(value)) {
      result.push(JSON.stringify(value, Object.getOwnPropertyNames(value)));
    } else if (isString(value)) {
      result.push(value);
    } else if (isNull(value)) {
      result.push("null");
    } else if (isUndefined(value)) {
      result.push("undefined");
    } else {
      result.push(JSON.stringify(value));
    }
  }

  return result;
};

const stringifyObjectValues = (input: Record<string, any>): Record<string, any> => {
  const result: Record<string, any> = {};

  for (const [key, value] of Object.entries(input)) {
    if (isObject(value)) {
      result[key] = stringifyObjectValues(value);
    } else if (isArray(value)) {
      result[key] = stringifyArrayValues(value);
    } else if (isDate(value)) {
      result[key] = value.toJSON();
    } else if (isError(value)) {
      result[key] = JSON.stringify(value, Object.getOwnPropertyNames(value));
    } else if (isString(value)) {
      result[key] = value;
    } else if (isNull(value)) {
      result[key] = "null";
    } else if (isUndefined(value)) {
      result[key] = "undefined";
    } else {
      result[key] = JSON.stringify(value);
    }
  }

  return result;
};

export const stringifyBlob = (object: Record<string, any>): string => {
  return JSON.stringify({
    json: stringifyObjectValues(object),
    meta: getMetaObject(object),
  });
};
