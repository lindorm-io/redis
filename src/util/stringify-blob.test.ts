import MockDate from "mockdate";
import { stringifyBlob } from "./stringify-blob";
import { parseBlob } from "./parse-blob";

MockDate.set("2020-01-01 08:00:00.000");

describe("stringifyBlob", () => {
  test("should create a text blob with json and meta values", () => {
    const blobObject: any = {
      array: [
        true,
        new Date(),
        12345,
        "string",
        {
          array: [true, 123, "arr", null, undefined],
          bool: false,
          date: new Date(),
          num: 123,
          str: "str",
        },
      ],
      boolean: true,
      date: new Date(),
      number: 12345,
      object: {
        a: "a",
        b: 1,
        c: null,
        d: true,
        e: [],
        f: undefined,
      },
      string: "string",
    };

    expect(stringifyBlob(blobObject)).toMatchSnapshot();
  });

  test("should create blob with errors", () => {
    const blob = stringifyBlob({
      error: new Error("error"),
    });

    expect(parseBlob(blob)).toStrictEqual({
      error: expect.any(Error),
    });
  });
});
