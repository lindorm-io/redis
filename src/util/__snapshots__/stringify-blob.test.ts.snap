// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringifyBlob should create a text blob with json and meta values 1`] = `"{\\"json\\":{\\"array\\":[\\"true\\",\\"2020-01-01T07:00:00.000Z\\",\\"12345\\",\\"string\\",{\\"array\\":[\\"true\\",\\"123\\",\\"arr\\",\\"null\\",\\"undefined\\"],\\"bool\\":\\"false\\",\\"date\\":\\"2020-01-01T07:00:00.000Z\\",\\"num\\":\\"123\\",\\"str\\":\\"str\\"}],\\"boolean\\":\\"true\\",\\"date\\":\\"2020-01-01T07:00:00.000Z\\",\\"number\\":\\"12345\\",\\"object\\":{\\"a\\":\\"a\\",\\"b\\":\\"1\\",\\"c\\":\\"null\\",\\"d\\":\\"true\\",\\"e\\":[],\\"f\\":\\"undefined\\"},\\"string\\":\\"string\\"},\\"meta\\":{\\"array\\":[\\"boolean\\",\\"date\\",\\"number\\",\\"string\\",{\\"array\\":[\\"boolean\\",\\"number\\",\\"string\\",\\"null\\",\\"undefined\\"],\\"bool\\":\\"boolean\\",\\"date\\":\\"date\\",\\"num\\":\\"number\\",\\"str\\":\\"string\\"}],\\"boolean\\":\\"boolean\\",\\"date\\":\\"date\\",\\"number\\":\\"number\\",\\"object\\":{\\"a\\":\\"string\\",\\"b\\":\\"number\\",\\"c\\":\\"null\\",\\"d\\":\\"boolean\\",\\"e\\":[],\\"f\\":\\"undefined\\"},\\"string\\":\\"string\\"}}"`;
